pallets/ddc-clusters/src/node_provider_auth.rs

```diff
@@ -52,19 +54,87 @@ where
 			false,
 		)
 		.result
-		.map_err(|_| NodeProviderAuthContractError::ContractCallFailed)?
+		.map_err(|_| NodeProviderAuthContractError::ContractCall)?
 		.data
 		.first()
 		.is_some_and(|x| *x == 1);
 
 		Ok(is_authorized)
 	}
 
+	pub fn deploy_contract(
+		&self,
+		caller_id: T::AccountId,
+	) -> Result<Self, NodeProviderAuthContractError> {
+		pub const CTOR_SELECTOR: [u8; 4] = hex!("9bae9d5e");
+
+		fn encode_constructor() -> Vec<u8> {
+			let mut call_data = CTOR_SELECTOR.to_vec();
+			let x = 0_u128;
+			for _ in 0..9 {
+				x.encode_to(&mut call_data);
+			}
+			call_data
+		}
+
+		// Load the contract code.
+		let wasm = &include_bytes!("./test_data/node_provider_auth_white_list.wasm")[..];
+		let _wasm_hash = <T as frame_system::Config>::Hashing::hash(wasm);
+		let contract_args = encode_constructor();
+
+		// Deploy the contract.
+		let contract_id = pallet_contracts::Pallet::<T>::bare_instantiate(
+			caller_id.clone(),
+			Default::default(),
+			EXTENSION_CALL_GAS_LIMIT,
+			None,
+			wasm.into(),
+			contract_args,
+			vec![],
+			false,
+		)
+		.result
+		.map_err(|_| NodeProviderAuthContractError::ContractDeploy)?
+		.account_id;
+
+		Ok(Self::new(contract_id, caller_id))
+	}
+
+	pub fn authorize_node(
+		&self,
+		node_pub_key: NodePubKey,
+	) -> Result<bool, NodeProviderAuthContractError> {
+		pub const ADD_DDC_NODE_SELECTOR: [u8; 4] = hex!("7a04093d");
+
+		let call_data = {
+			// is_authorized(node_provider: AccountId, node: Vec<u8>, node_variant: u8) -> bool
+			let args: ([u8; 4], Vec<u8>) =
+				(ADD_DDC_NODE_SELECTOR, node_pub_key.encode()[1..].to_vec());
+			args.encode()
+		};
+
+		let _ = pallet_contracts::Pallet::<T>::bare_call(
+			self.caller_id.clone(),
+			self.contract_id.clone(),
+			Default::default(),
+			EXTENSION_CALL_GAS_LIMIT,
+			None,
+			call_data,
+			false,
+		)
+		.result
+		.map_err(|_| NodeProviderAuthContractError::NodeAuthorization)?;
+
+		Ok(true)
+	}
+
 	pub fn new(contract_id: T::AccountId, caller_id: T::AccountId) -> Self {
 		Self { contract_id, caller_id }
 	}
 }
 
 pub enum NodeProviderAuthContractError {
-	ContractCallFailed,
+	ContractCall,
```

Comment:
I would leave it as `ContractCallFailed`, and add `ContractDeployFailed`, `NodeAuthorizationUnsucceeded` to make error type more descriptive

Reply:
I got an error from linter, as suffix was the same, but will find a way around

Reply:
Fixed